### PR TITLE
KEYCLOAK-1267 Add dedicated SSO timeouts for Remember-Me

### PR DESCRIPTION
**Pull request by thomasdarimont**
Previously remember-me sessions where tied to the SSO max session
timeout which could lead to unexpected early session timeouts.
We now allow SSO timeouts to be configured separately for sessions
with enabled remember-me.
This enables users to opt-in for longer session timeouts.

SSO Session Timeouts for RememberMe can now be configured in the
tokens tab in the realm admin console. Max SSO timeouts for Remember-me
need to be equal or greater than normal max SSO timeouts.

Note that most of the work for this PR was done by @cfsnyder and
is based on his earlier PR https://github.com/keycloak/keycloak/pull/2338.
I just adapted his code to the latest Keycloak API changes.

Signed-off-by: Thomas Darimont thomas.darimont@gmail.com

### CODE REVIEW

**Comment by cfsnyder** [comment]
Awesome @thomasdarimont, thanks for following up and updating that one!

File: model/jpa/src/main/resources/META-INF/jpa-changelog-2.2.0.xml [hunk 1]
```diff
@@ -57,5 +57,13 @@
 
     </changeSet>
 
+    <changeSet author="keycloak" id="2.2.0.1">
```
Comment (stianst) [inline model/jpa/src/main/resources/META-INF/jpa-changelog-2.2.0.xml]: Please don't introduce extra changeSet for a specific version, we only want one changeset per-version. Just append to existing changeSet element if there is one.

File: model/jpa/src/main/resources/META-INF/jpa-changelog-2.2.0.xml [hunk 2]
```diff
@@ -57,5 +57,13 @@
...
+        <addColumn tableName="REALM">
+            <column name="SSO_MAX_LIFESPAN_REMEMBER_ME" type="INT" defaultValueNumeric="3600"/>
+        </addColumn>
+
+        <!-- intialize the SSO_MAX_LIFESPAN_REMEMBER_ME = SSO_MAX_LIFESPAN for all pre-existing realms -->
+        <sql>update REALM set SSO_MAX_LIFESPAN_REMEMBER_ME = SSO_MAX_LIFESPAN where SSO_MAX_LIFESPAN is not null</sql>
```
Comment (stianst) [inline model/jpa/src/main/resources/META-INF/jpa-changelog-2.2.0.xml]: Would be simpler to allow null and use ssoSessionMaxLifespan if not specified. Having SQL in these files worries me for compatibility reasons (in general, I think this particular SQL is fine though).

**Review by stianst — COMMENTED** [review]
Sorry that it took so long to get back to you on this one. Looks OK, but I've added a couple comments inline. We need 'ssoSessionIdleTimeoutRememberMe' as well. This gives more flexibility in how to configure things. For example remember me allows cookie to last 365 days, but still needs a user to use the session at least every 30 days, while a non-remember me session is 30 days and idle 1 day.

**Comment by thomasdarimont** [comment]
FYI I updated the PR with an additional idle timeout for sso remember me. 
Would you mind having a look @stianst and @cfsnyder ?

File: model/jpa/src/main/resources/META-INF/jpa-changelog-2.3.0.xml [hunk 3]
```diff
@@ -36,6 +36,16 @@
          <addColumn tableName="IDENTITY_PROVIDER">
              <column name="PROVIDER_DISPLAY_NAME" type="VARCHAR(255)"></column>
          </addColumn>
+
+         <addColumn tableName="REALM">
+             <column name="SSO_MAX_LIFESPAN_REMEMBER_ME" type="INT" defaultValueNumeric="3600"/>
```
Comment (thomasdarimont) [inline model/jpa/src/main/resources/META-INF/jpa-changelog-2.3.0.xml]: should the default be 1 hour or 10 hours (as in the examples above e.g.: basicauthrealm.json)?

File: model/jpa/src/main/resources/META-INF/jpa-changelog-2.3.0.xml [hunk 4]
```diff
@@ -36,6 +36,16 @@
...
+             <column name="SSO_IDLE_TIMEOUT_REMEMBER_ME" type="INT" defaultValueNumeric="3600"/>
```
Comment (thomasdarimont) [inline model/jpa/src/main/resources/META-INF/jpa-changelog-2.3.0.xml]: I'm not sure what a better default would be here...

File: model/jpa/src/main/resources/META-INF/jpa-changelog-2.3.0.xml [hunk 5]
```diff
@@ -36,6 +36,16 @@
          <addColumn tableName="IDENTITY_PROVIDER">
              <column name="PROVIDER_DISPLAY_NAME" type="VARCHAR(255)"></column>
          </addColumn>
+
+         <addColumn tableName="REALM">
+             <column name="SSO_MAX_LIFESPAN_REMEMBER_ME" type="INT" defaultValueNumeric="36000"/>
```
Comment (thomasdarimont) [inline model/jpa/src/main/resources/META-INF/jpa-changelog-2.3.0.xml]: set this to 10 hours by default - can anyone think of a better default?

File: services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java [hunk 6]
```diff
@@ -97,8 +97,11 @@ public static boolean isSessionValid(RealmModel realm, UserSessionModel userSess
             return false;
         }
         int currentTime = Time.currentTime();
-        int max = userSession.getStarted() + realm.getSsoSessionMaxLifespan();
-        return userSession.getLastSessionRefresh() + realm.getSsoSessionIdleTimeout() > currentTime && max > currentTime;
+        boolean sessionMaxOk = (userSession.getStarted() + (userSession.isRememberMe() ? realm.getSsoSessionMaxLifespanRememberMe() : realm.getSsoSessionMaxLifespan())) > currentTime;
+        int maxRememberMeIdleSessionAge = userSession.getLastSessionRefresh() + realm.getSsoSessionIdleTimeoutRememberMe();
+        int maxSsoSessionAge = userSession.getLastSessionRefresh() + realm.getSsoSessionIdleTimeout();
+        boolean sessionIdleOk = (userSession.isRememberMe() && maxRememberMeIdleSessionAge > currentTime) || maxSsoSessionAge > currentTime;
```
Comment (thomasdarimont) [inline services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java]: Note that  I now take into account the `ssoSessionIdleTimeoutRememberMe`

File: model/jpa/src/main/resources/META-INF/jpa-changelog-2.2.0.xml (hunk 1 above)
Comment (thomasdarimont) [inline model/jpa/src/main/resources/META-INF/jpa-changelog-2.2.0.xml]: Okay, I moved the changes to the current changelog.

File: examples/basic-auth/basicauthrealm.json [hunk 7]
```diff
@@ -6,6 +6,8 @@
     "accessCodeLifespanUserAction": 300,
     "ssoSessionIdleTimeout": 600,
     "ssoSessionMaxLifespan": 36000,
+    "ssoSessionIdleTimeoutRememberMe": 36000,
```
Comment (stianst) [inline examples/basic-auth/basicauthrealm.json]: Don't add to existing realm files. That should not be required.

File: examples/cors/cors-realm.json [hunk 8]
```diff
@@ -6,6 +6,8 @@
     "accessCodeLifespanUserAction": 300,
     "ssoSessionIdleTimeout": 600,
     "ssoSessionMaxLifespan": 36000,
+    "ssoSessionIdleTimeoutRememberMe": 36000,
```
Comment (stianst) [inline examples/cors/cors-realm.json]: Don't add to existing realm files. That should not be required.

File: examples/demo-template/testrealm.json [hunk 9]
```diff
@@ -6,6 +6,8 @@
     "accessCodeLifespanUserAction": 300,
     "ssoSessionIdleTimeout": 600,
     "ssoSessionMaxLifespan": 36000,
+    "ssoSessionIdleTimeoutRememberMe": 36000,
```
Comment (stianst) [inline examples/demo-template/testrealm.json]: Don't add to existing realm files. That should not be required.

File: examples/fuse/testrealm.json [hunk 10]
```diff
@@ -6,6 +6,8 @@
     "accessCodeLifespanUserAction": 300,
     "ssoSessionIdleTimeout": 600,
     "ssoSessionMaxLifespan": 36000,
+    "ssoSessionIdleTimeoutRememberMe": 36000,
```
Comment (stianst) [inline examples/fuse/testrealm.json]: Don't add to existing realm files. That should not be required.

File: model/jpa/src/main/resources/META-INF/jpa-changelog-2.2.0.xml [hunk 11]
```diff
@@ -21,7 +21,7 @@
     <changeSet author="bburke@redhat.com" id="2.2.0">
 
         <addColumn tableName="ADMIN_EVENT_ENTITY">
-            <column name="RESOURCE_TYPE" type="VARCHAR(64)"></column>
+            <column name="RESOURCE_TYPE" type="VARCHAR(64)"/>
```
Comment (stianst) [inline model/jpa/src/main/resources/META-INF/jpa-changelog-2.2.0.xml]: Don't change code that is not relevant to your changes

File: model/jpa/src/main/resources/META-INF/jpa-changelog-2.3.0.xml (hunk 5 above)
Comment (stianst) [inline model/jpa/src/main/resources/META-INF/jpa-changelog-2.3.0.xml]: I don't like having a default for this. Rather values for remember me should be optional and it should use non-remember me values if not specified.

File: model/jpa/src/main/resources/META-INF/jpa-changelog-2.3.0.xml [hunk 12]
```diff
@@ -36,6 +36,16 @@
...
+             <column name="SSO_IDLE_TIMEOUT_REMEMBER_ME" type="INT" defaultValueNumeric="36000"/>
+         </addColumn>
+
+         <!-- intialize the SSO_MAX_LIFESPAN_REMEMBER_ME = SSO_MAX_LIFESPAN for all pre-existing realms -->
+         <sql>update REALM set SSO_MAX_LIFESPAN_REMEMBER_ME = SSO_MAX_LIFESPAN where SSO_MAX_LIFESPAN is not null</sql>
```
Comment (stianst) [inline model/jpa/src/main/resources/META-INF/jpa-changelog-2.3.0.xml]: This is doomed to fail. See above though. It think this should not be set by default and be optional to specify.

File: server-spi/src/main/java/org/keycloak/models/utils/RepresentationToModel.java [hunk 13]
```diff
@@ -162,6 +162,12 @@ public static void importRealm(KeycloakSession session, RealmRepresentation rep,
         else newRealm.setSsoSessionIdleTimeout(1800);
         if (rep.getSsoSessionMaxLifespan() != null) newRealm.setSsoSessionMaxLifespan(rep.getSsoSessionMaxLifespan());
         else newRealm.setSsoSessionMaxLifespan(36000);
+
+        if (rep.getSsoSessionIdleTimeoutRememberMe() != null) newRealm.setSsoSessionIdleTimeoutRememberMe(rep.getSsoSessionIdleTimeoutRememberMe());
+        else newRealm.setSsoSessionIdleTimeoutRememberMe(36000);
```
Comment (stianst) [inline server-spi/src/main/java/org/keycloak/models/utils/RepresentationToModel.java]: Same as above. It should not be set by default and use non-remember me values if not specified explicitly.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/composites/CompositeRoleTest.java [hunk 14]
```diff
@@ -56,6 +56,8 @@ public void addTestRealms(List<RealmRepresentation> testRealms) {
                 .ssoSessionIdleTimeout(3000)
                 .accessTokenLifespan(10000)
                 .ssoSessionMaxLifespan(10000)
+                .ssoSessionMaxLifespanRememberMe(10000)
```
Comment (stianst) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/composites/CompositeRoleTest.java]: Shouldn't be required to set this in existing tests. This test has nothing to do with your changes.

File: testsuite/integration-arquillian/tests/base/src/test/resources/testrealm.json [hunk 15]
```diff
@@ -6,6 +6,7 @@
   "registrationAllowed": true,
   "resetPasswordAllowed": true,
   "editUsernameAllowed" : true,
+  "rememberMe" : false,
```
Comment (stianst) [inline testsuite/integration-arquillian/tests/base/src/test/resources/testrealm.json]: Why are you changing this?

File: testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/AbstractConsoleTest.java [hunk 16]
```diff
@@ -77,25 +80,35 @@ public void beforeConsoleTest() {
     }
 
     public void loginToMasterRealmAdminConsoleAs(UserRepresentation user) {
-        loginToAdminConsoleAs(adminConsolePage, loginPage, user);
+        loginToAdminConsoleAs(adminConsolePage, loginPage, user, false);
```
Comment (stianst) [inline testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/AbstractConsoleTest.java]: I don't understand the point in adding this. This is not the correct place to test remember me.

File: testsuite/integration-arquillian/tests/other/mod_auth_mellon/src/test/resources/mellon-realm.json [hunk 17]
```diff
@@ -7,6 +7,8 @@
   "accessTokenLifespanForImplicitFlow" : 900,
   "ssoSessionIdleTimeout" : 2,
   "ssoSessionMaxLifespan" : 36000,
+  "ssoSessionIdleTimeoutRememberMe" : 36000,
```
Comment (stianst) [inline testsuite/integration-arquillian/tests/other/mod_auth_mellon/src/test/resources/mellon-realm.json]: Please don't add to irrelevant tests

File: testsuite/integration/src/test/java/org/keycloak/testsuite/adapter/AdapterTest.java [hunk 18]
```diff
@@ -126,6 +126,11 @@ public void testLoginSSOMax() throws Exception {
         testStrategy.testLoginSSOMax();
     }
 
+    @Test
```
Comment (stianst) [inline testsuite/integration/src/test/java/org/keycloak/testsuite/adapter/AdapterTest.java]: Adapter tests are not the correct place to test this. Rather use LoginTest or something else from the base tests inside the new Arquillian testsuite. Please remove all changes to the adapter tests inside testsuite/integration

**Review by stianst — COMMENTED** [review]
A few comments inline. In general I think it would be better to use non-remember me values by default and optionally allow setting different values for remember-me.

**Pull request closed without merging**